### ISSUE TITLE
.abnf grammar

### ISSUE THREAD

**Issue by venkatd** [opencypher/openCypher#30]
Hi, I'm working on building a pretty deep neo4j integration with Elixir. I'm pretty new to grammars and I have only found parsers that support anbf grammars. I'm not familiar with the difference between abnf and ebnf.

Is this something that is easy to add? It looks like you are generating the .enbf from an xml file.

Thanks!

**Comment by Mats-SX** [comment]
Hello @venkatd and thanks for your question.

The thing with grammars is that each parser or tool seems to have their own interpretation of BNF-like syntax. We went with EBNF because it has an ISO standard (ISO14977), but we do not expect that the generated artifact (the EBNF grammar) will be readily consumed in its raw form. Rather we'd expect some tool to have to rewrite it into the target flavour of EBNF syntax, and I think it would be possible to do so for ABNF as well.

It would probably be possible to add a custom ABNF-writer, however, which perhaps is what you really were after. You're correct that we are generating the grammars from an XML specification. This XML specification is what we regard as the source of truth (although it is not currently complete), rather than the generated BNF grammars. We do currently have two such writers, one that uses ISO14977 format, and one that uses Antlrv4 format (http://www.antlr.org/). I don't see any immediate blockers to adding another subclass of `BnfWriter`, `AbnfWriter` or similar, that would output the grammar in the ABNF format. From the little I've gathered on ABNF so far, it seems to be similar enough to EBNF that the current ISO14977 implementation could be of good use in such an effort.

As for our own plans, we do not currently intend on adding more output formats, but we welcome both the discussion and possible contributions, if you are interested in implementing something like this yourself.

All the best
Mats
